multistream-2.0/spec.md

```diff
@@ -187,6 +187,75 @@ Importantly: When we switch to multistream 2.0, we'll tag the connection (and
 any sub connections) with the multistream version. This way, we never have to do
 this again.
 
+## TCP Simultaneous Open
+
+As noted in the [retrospective](retrospective.md), multistream 1.0 doesn't
+provide any way to distinguish between the initiator and the receiver of a
+stream. Multistream 2.0 doesn't either but it *does* allow us to handle the TCP
+Simultaneous Open case without needing any additional round-trips in the fast
+path.
+
+To make this work, we need a new protocol: "duplex-stream" (or whatever we want
+to call it). This protocol allows one to bind two unidirectional streams
+together into a single bidirectional stream.
+
+### Protocol
+
+The protocol is:
+
+1. The side that wants to be the "initiator" of a duplex stream sends a
+   "initiate stream ID" message (where ID is randomly generated 256 bit number).
+2. The receiver sends back "receive stream ID" on a different unidirectional stream.
+3. The two streams are now joined.
+
+More specifically,
+
+1. The initiator sends: `<multistream/multicodec><duplex-stream>0<32 bytes of randomness>`
+1. The receiver sends: `<multistream/multicodec><duplex-stream>1<the same 32 bytes of randomness>`
+
+If we end up in a situation where both peers want to be the initiator of a
+single pair of unidirectional streams, the peer that picks the *lower* random ID
+should back off and act as the receiver.
+
+### Usage
+
+We treat each new TCP connection as a pair of unidirectional streams and use
+this protocol to bind them together.
```

Comment:
I think we need full detail on the wire for this, all the way up to secio negotiation (both cases).